File: python/src/find_itch_games/receipt.py
"""Reading the receipts itch leaves in install folders.

Every folder itch installs a game into gets a ``.itch/receipt.json.gz``
describing what was put there. It is the on-disk source of truth: it survives a
database reset, and its presence is what distinguishes an itch game folder from
any other directory.
"""

from __future__ import annotations

import gzip
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .types import Build, Game, Upload
from .utils import read_subdirectories

__all__ = [
    "RECEIPT_DIR",
    "FoundReceipt",
    "LegacyReceiptInfo",
    "Receipt",
    "find_receipts",
    "get_legacy_receipt_path",
    "get_receipt_path",
    "has_receipt",
    "read_receipt",
]

#: The metadata directory itch keeps inside an install folder.
RECEIPT_DIR = ".itch"

#: itch's staging area, never a game folder. Matches butler's own scan.
_STAGING_FOLDER = "downloads"


@dataclass
class LegacyReceiptInfo:
    """What the pre-v23 uncompressed ``.itch/receipt.json`` recorded.

    Mirrors butler's ``legacyReceipt``/``legacyCave`` structs, which it still
    reads when scanning install locations. There is no game title or url in
    that format, only ids.
    """

    game_id: int
    cave_id: str | None = None
    upload_id: int | None = None
    build_id: int | None = None
    install_location: str | None = None
    install_folder: str | None = None
    path_scheme: int | None = None
    seconds_run: float | None = None
    installed_at: Any | None = None
    last_touched: float | None = None


@dataclass
class Receipt:
    """itch's record of what it installed into a folder.

    Mirrors ``bfs.Receipt`` in ``github.com/itchio/hush``.
    """

    game: Game
    upload: Upload | None = None
    build: Build | None = None
    files: list[str] = field(default_factory=list)
    installer_name: str | None = None
    #: Set when the receipt came from the pre-v23 format.
    legacy: LegacyReceiptInfo | None = None


def get_receipt_path(install_path: Path | str) -> Path:
    """The current receipt path: ``<install_path>/.itch/receipt.json.gz``."""
    return Path(install_path) / RECEIPT_DIR / "receipt.json.gz"


def get_legacy_receipt_path(install_path: Path | str) -> Path:
    """The pre-v23 receipt path: ``<install_path>/.itch/receipt.json``."""
    return Path(install_path) / RECEIPT_DIR / "receipt.json"


def _receipt_from_modern(data: dict[str, Any]) -> Receipt | None:
    game = Game.from_dict(data.get("game"))
    if game is None or not isinstance(data.get("game", {}).get("id"), int):
        return None
    return Receipt(
        game=game,
        upload=Upload.from_dict(data.get("upload")),
        build=Build.from_dict(data.get("build")),
        files=list(data.get("files") or []),
        installer_name=data.get("installerName"),
    )


def _read_modern(receipt_path: Path) -> Receipt | None:
    try:
        raw = receipt_path.read_bytes()
    except OSError:
        return None
    try:
        # butler always gzips, but decide from the magic bytes rather than the
        # extension so a hand-edited receipt still reads.
        text = gzip.decompress(raw) if raw[:2] == b"\x1f\x8b" else raw
        data = json.loads(text)
    except (OSError, ValueError):
        return None
    return _receipt_from_modern(data) if isinstance(data, dict) else None


def _read_legacy(receipt_path: Path) -> Receipt | None:
    try:
        data = json.loads(receipt_path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None
    if not isinstance(data, dict):
        return None

    # Some builds wrote the modern shape uncompressed to this path.
    modern = _receipt_from_modern(data)
    if modern is not None:
        return modern

    cave = data.get("cave")
    if not isinstance(cave, dict) or not isinstance(cave.get("gameId"), int):
        return None

    build_id = cave.get("buildId") or None
    return Receipt(
        game=Game(id=cave["gameId"]),
        upload=Upload(id=cave["uploadId"]) if isinstance(cave.get("uploadId"), int) else None,
        build=Build(id=build_id) if build_id else None,
        files=list(data.get("files") or []),
        legacy=LegacyReceiptInfo(
            game_id=cave["gameId"],
            cave_id=cave.get("id"),
            upload_id=cave.get("uploadId"),
            build_id=build_id,
            install_location=cave.get("installLocation"),
            install_folder=cave.get("installFolder"),
            path_scheme=cave.get("pathScheme"),
            seconds_run=cave.get("secondsRun"),
            installed_at=cave.get("installedAt"),
            last_touched=cave.get("lastTouched"),
        ),
    )


def read_receipt(install_path: Path | str) -> Receipt | None:
    """Reads the receipt itch left inside an install folder.

    Falls back to the pre-v23 ``receipt.json`` format, normalized into the same
    shape with its ids under :attr:`Receipt.legacy`. Returns ``None`` if the
    folder has no readable receipt.
    """
    modern = _read_modern(get_receipt_path(install_path))
    if modern is not None:
        return modern
    return _read_legacy(get_legacy_receipt_path(install_path))


def has_receipt(install_path: Path | str) -> bool:
    """Whether an install folder carries an itch receipt."""
    return read_receipt(install_path) is not None


@dataclass
class FoundReceipt:
    """A receipt found by scanning an install location, and where it was."""

    path: Path
    install_folder_name: str
    receipt: Receipt


def find_receipts(library_path: Path | str) -> list[FoundReceipt]:
    """Scans the immediate subdirectories of an install location for receipts.

    Follows the same rules as butler's own install-location scan: skip the
    ``downloads`` staging folder, require a ``.itch`` directory, then read the
    receipt.
    """
    library = Path(library_path)
    if not library.is_dir():
        return []

    found: list[FoundReceipt] = []
    for name in read_subdirectories(library):
        if name == _STAGING_FOLDER:
            continue
        install_path = library / name
        if not (install_path / RECEIPT_DIR).is_dir():
            continue
        receipt = read_receipt(install_path)
        if receipt is not None:
            found.append(FoundReceipt(path=install_path, install_folder_name=name, receipt=receipt))
    return found

File: node/src/receipt.ts
/**
 * Reading the receipts itch leaves in install folders.
 *
 * Every folder itch installs a game into gets a `.itch/receipt.json.gz`
 * describing what was put there. It is the on-disk source of truth: it
 * survives a database reset, and its presence is what distinguishes an itch
 * game folder from any other directory.
 *
 * @module
 */
import * as fs from "node:fs/promises";
import * as path from "node:path";
import { gunzip } from "node:zlib";
import { promisify } from "node:util";

import type { IItchBuild, IItchGame, IItchUpload } from "./types.js";
import { isDirectory, parseJson, readSubdirectories } from "./utils.js";

const gunzipAsync = promisify(gunzip);

/**
 * itch writes one of these into every folder it installs a game into.
 *
 * Mirrors `bfs.Receipt` in `github.com/itchio/hush`.
 */
export interface IItchReceipt {
  /** The itch.io game installed at this location. */
  game: IItchGame;
  /** The itch.io upload installed at this location. */
  upload?: IItchUpload;
  /** The itch.io build installed here. Absent for non-wharf uploads. */
  build?: IItchBuild;
  /** Installed files, as slash-separated paths relative to the folder. */
  files?: string[];
  /** The installer used, e.g. `archive`, `msi`, `inno`. */
  installerName?: string;
  /**
   * Set when the receipt came from the pre-v23 `.itch/receipt.json` format,
   * which recorded ids but no game metadata.
   */
  legacy?: IItchLegacyReceiptInfo;
}

/**
 * What the pre-v23 uncompressed `.itch/receipt.json` recorded.
 *
 * Mirrors butler's `legacyReceipt`/`legacyCave` structs, which it still reads
 * when scanning install locations. There is no game title or url in this
 * format, only ids.
 */
export interface IItchLegacyReceiptInfo {
  caveId?: string;
  gameId: number;
  uploadId?: number;
  buildId?: number;
  /** Id of the install location, as the old client recorded it. */
  installLocation?: string;
  installFolder?: string;
  /** `2` is the only scheme butler will import. */
  pathScheme?: number;
  secondsRun?: number;
  /** A unix timestamp or an RFC3339 string, depending on the client version. */
  installedAt?: number | string;
  lastTouched?: number;
}

/** The metadata directory itch keeps inside an install folder. */
export const RECEIPT_DIR = ".itch";

/**
 * The current receipt path: `<installPath>/.itch/receipt.json.gz`.
 *
 * @param installPath - A game's install folder.
 */
export function getReceiptPath(installPath: string): string {
  return path.join(installPath, RECEIPT_DIR, "receipt.json.gz");
}

/**
 * The pre-v23 receipt path: `<installPath>/.itch/receipt.json`.
 *
 * @param installPath - A game's install folder.
 */
export function getLegacyReceiptPath(installPath: string): string {
  return path.join(installPath, RECEIPT_DIR, "receipt.json");
}

/**
 * Where {@link readReceipt} looks, in the order butler looks.
 *
 * @param installPath - A game's install folder.
 */
export function getReceiptPaths(installPath: string): string[] {
  return [getReceiptPath(installPath), getLegacyReceiptPath(installPath)];
}

interface ILegacyReceiptFile {
  cave?: {
    id?: string;
    gameId?: number;
    uploadId?: number;
    buildId?: number;
    installLocation?: string;
    installFolder?: string;
    pathScheme?: number;
    secondsRun?: number;
    installedAt?: number | string;
    lastTouched?: number;
  };
  files?: string[];
}

/**
 * Reads the receipt itch left inside an install folder.
 *
 * Falls back to the pre-v23 `receipt.json` format, which is normalized into
 * the same shape with the ids it carries under `legacy`.
 *
 * @param installPath - A game's install folder.
 * @returns The receipt, or `null` if the folder has none.
 */
export async function readReceipt(installPath: string): Promise<IItchReceipt | null> {
  const modern = await readModernReceipt(getReceiptPath(installPath));
  if (modern !== null) return modern;
  return readLegacyReceipt(getLegacyReceiptPath(installPath));
}

async function readModernReceipt(receiptPath: string): Promise<IItchReceipt | null> {
  let raw: Buffer;
  try {
    raw = await fs.readFile(receiptPath);
  } catch {
    return null;
  }

  // butler always gzips, but decide from the magic bytes rather than the
  // extension so a hand-edited receipt still reads.
  const isGzipped = raw[0] === 0x1f && raw[1] === 0x8b;
  let text: string;
  try {
    text = isGzipped ? (await gunzipAsync(raw)).toString("utf8") : raw.toString("utf8");
  } catch {
    return null;
  }

  const receipt = parseJson<IItchReceipt>(text);
  return typeof receipt?.game?.id === "number" ? receipt : null;
}

async function readLegacyReceipt(receiptPath: string): Promise<IItchReceipt | null> {
  let text: string;
  try {
    text = await fs.readFile(receiptPath, "utf8");
  } catch {
    return null;
  }

  const parsed = parseJson<ILegacyReceiptFile | IItchReceipt>(text);
  if (parsed === null) return null;

  // Some builds wrote the modern shape uncompressed to this path.
  if (typeof (parsed as IItchReceipt).game?.id === "number") return parsed as IItchReceipt;

  const cave = (parsed as ILegacyReceiptFile).cave;
  if (typeof cave?.gameId !== "number") return null;

  return {
    game: { id: cave.gameId },
    ...(typeof cave.uploadId === "number" ? { upload: { id: cave.uploadId } } : {}),
    ...(typeof cave.buildId === "number" && cave.buildId !== 0
      ? { build: { id: cave.buildId } }
      : {}),
    files: (parsed as ILegacyReceiptFile).files ?? [],
    legacy: {
      caveId: cave.id,
      gameId: cave.gameId,
      uploadId: cave.uploadId,
      buildId: cave.buildId || undefined,
      installLocation: cave.installLocation,
      installFolder: cave.installFolder,
      pathScheme: cave.pathScheme,
      secondsRun: cave.secondsRun,
      installedAt: cave.installedAt,
      lastTouched: cave.lastTouched,
    },
  };
}

/**
 * Whether an install folder carries an itch receipt.
 *
 * @param installPath - A game's install folder.
 */
export async function hasReceipt(installPath: string): Promise<boolean> {
  return (await readReceipt(installPath)) !== null;
}

/** A receipt found by scanning an install location, with where it was found. */
export interface IFoundReceipt {
  /** Absolute path to the install folder. */
  path: string;
  /** The folder's own name, which is itch's `installFolderName`. */
  installFolderName: string;
  receipt: IItchReceipt;
}

/**
 * Scans the immediate subdirectories of an install location for receipts.
 *
 * This follows the same rules as butler's own install-location scan: skip the
 * `downloads` staging folder, require a `.itch` directory, then read the
 * receipt. It is the on-disk source of truth -- it finds games butler.db has
 * forgotten, and it ignores unrelated folders, which matters because an
 * install location is often a directory the user keeps other things in.
 *
 * @param libraryPath - An install location's path.
 */
export async function findReceipts(libraryPath: string): Promise<IFoundReceipt[]> {
  if (!(await isDirectory(libraryPath))) return [];

  const names = await readSubdirectories(libraryPath);
  const found = await Promise.all(
    names.map(async (installFolderName): Promise<IFoundReceipt | null> => {
      // itch's own staging area, never a game folder.
      if (installFolderName === "downloads") return null;

      const installPath = path.join(libraryPath, installFolderName);
      if (!(await isDirectory(path.join(installPath, RECEIPT_DIR)))) return null;

      const receipt = await readReceipt(installPath);
      return receipt === null ? null : { path: installPath, installFolderName, receipt };
    }),
  );

  return found.filter((entry): entry is IFoundReceipt => entry !== null);
}
